pancancer_evaluation/utilities/feature_utilities.py

```diff
@@ -0,0 +1,91 @@
+"""
+Utilities for testing feature selection methods.
+
+subset_by_feature_weights() is the main external function, and the
+internal functions implement the different methods for ranking/selecting
+features.
+"""
+import warnings
+
+import numpy as np
+import pandas as pd
+import scipy.stats as st
+from sklearn.feature_selection import f_classif
+
+def subset_by_feature_weights(X_train_df,
+                              X_test_df,
+                              feature_selection_method,
+                              gene_features,
+                              y_df,
+                              num_features):
+
+    gene_feature_names = X_train_df.columns.values[gene_features]
+    non_gene_features = X_train_df.columns.values[~gene_features]
+    feature_df = _get_cancer_type_f_statistics(
+        X_train_df.reindex(gene_feature_names, axis='columns'), y_df
+    )
+    weights_df = _generate_feature_weights(feature_df, feature_selection_method)
+
+    # for MAD we want to take the features with the *lowest* variance
+    # between cancer types
+    # for all others we want to take the features with the highest weight
+    ascending = (feature_selection_method == 'mad_f_test')
+    top_features = weights_df.sort_values(ascending=ascending).index[:num_features]
+    valid_features = np.concatenate((top_features, non_gene_features))
+
+    # y_df will get reindexed later to match the X indices
+    X_train_df = X_train_df.reindex(valid_features, axis='columns')
+    X_test_df = X_test_df.reindex(valid_features, axis='columns')
+    gene_features = np.concatenate((
+        np.ones(top_features.shape[0]).astype('bool'),
+        np.zeros(non_gene_features.shape[0]).astype('bool')
+    ))
+
+    return X_train_df, X_test_df, gene_features
+
+
+def _get_cancer_type_f_statistics(X_train_df, y_df):
```

Comment:
This may be blatantly obvious to anyone with more Python training than me, but what's the significance of putting the underscore in front of some functions but not others? You're not planning to call those directly in your main scripts?

Reply:
Yeah, it's the Python "internal method" convention (Python doesn't really support private/internal methods, so it's more like "please don't use this function outside of this file, even though you technically can"). I mostly use it when I want to write a complicated helper function but not clean it up or make it general enough to be used elsewhere, haha.

More detail here, if you're interested: https://stackoverflow.com/a/6930223